src/uri.jl

```diff
@@ -32,38 +22,44 @@ For efficiency, the internal representation is stored as a set of offsets and le
 To access and return these components as strings, use the various accessor methods:
   * `HTTP.scheme`: returns the scheme (if any) associated with the uri
   * `HTTP.userinfo`: returns the userinfo (if any) associated with the uri
-  * `HTTP.hostname`: returns the hostname only of the uri
+  * `HTTP.host`: returns the host only of the uri
   * `HTTP.port`: returns the port of the uri; will return "80" or "443" by default if the scheme is "http" or "https", respectively
-  * `HTTP.host`: returns the "hostname:port" combination; if the port is not provided or is the default port for the uri scheme, it will be omitted
+  * `HTTP.hostport`: returns the "host:port" combination; if the port is not provided or is the default port for the uri scheme, it will be omitted
   * `HTTP.path`: returns the path for a uri
   * `HTTP.query`: returns the query for a uri
   * `HTTP.fragment`: returns the fragment for a uri
   * `HTTP.resource`: returns the path-query-fragment combination
 """
 struct URI
-    data::Vector{UInt8}
-    offsets::NTuple{7, Offset}
+    uri::String
```

Comment:
Not a big fan of this change; the existing `URI` definition is very concise and performant and the "getters" are great for accessing necessary fields. with getfield overloading, they could even be made more convenient

Reply:
`SubString` fields that all reference a single master `String` should be equal if not superior in performance to the previous scheme of indexing into an array of offsets and using one of the offsets to index into a byte array and then creating a new String object from that.

But mostly this is about the principle of least surprise. When I switched from the old URI package to HTTP.jl I remember the first time I typed `uri.` in the REPL and pressed `<TAB><TAB>`:
```julia
julia> uri = HTTP.URI("http://foo.bar/fum")
HTTP.URI("http://foo.bar/fum")

julia> uri.<TAB><TAB>
data    offsets
```
vs
```julia
julia> u.
fragment host      path      port      query     scheme    uri       userinfo
```